entrypoint.rb

```diff
@@ -1,31 +1,41 @@
-require 'openai'
-require 'dotenv/load'
-require 'json'
-require 'pry'
-require_relative 'app/open_ai_client'
-require_relative 'app/submission'
-require_relative 'app/pupilfirst_api'
+require "openai"
+require "dotenv/load"
+require "pry"
+require_relative "app/open_ai_client"
+require_relative "app/submission"
+require_relative "app/pupilfirst_api"
+require_relative "app/reviewer"
 
 OpenAI.configure do |config|
-    config.access_token = ENV.fetch("OPENAI_ACCESS_TOKEN")
-    config.organization_id = ENV.fetch("OPENAI_ORGANIZATION_ID", "") # Optional.
+  config.access_token = ENV.fetch("OPENAI_ACCESS_TOKEN")
+  config.organization_id = ENV.fetch("OPENAI_ORGANIZATION_ID", "") # Optional.
 end
 
 def generate_response
-    @generate_response ||=
-    begin
-        JSON.parse(OpenAIClient.new.ask)
-    rescue => exception
-        {
-            status: "skip",
-            feedback: "Failure message"
-        }
-    end
+  retries = 0
+  generate_response = OpenAIClient.new.ask
+
+  case generate_response[:function_name]
+  when "create_grading"
+    PupilfirstAPI::Grader.new.grade(generate_response[:args])
+  when "create_feedback"
+    PupilfirstAPI::Grader.new.add_feedback(generate_response[:args])
+  else
+    raise "Unknown response from OpenAI: #{generate_response}"
+  end
+rescue => e
+  retries += 1
+  if retries <= 1
+    puts "Attempting retry #{retries} due to error: #{e.message}."
+    retry
+  else
+    puts "Max retries reached. Exiting."
+    raise e
+  end
 end
 
-if ENV.fetch("SKIP_GRADING", "false") == "true"
-    PupilfirstAPI::Grader.new.add_feedback(generate_response)
+if ENV.fetch("TEST_MODE", "false") == "true"
+  puts "Running in test mode. Skipping API calls."
 else
-    puts "Grading..."
-    PupilfirstAPI::Grader.new.grade(generate_response)
+  generate_response
```

Comment:
Isn't setting `TEST_MODE` to `true` supposed to run everything _except_ the API calls? Here, it looks like setting `TEST_MODE` to true will cause the action to... do nothing...? The call to `generate_response` is what causes `PupilfirstAPI::Grader` class to get used.

The `Grader` class, by the way, handles `TEST_MODE`, so I'm not sure what's going on here. 🤔

Reply:
This is discussed in the call.  Now, we have set the key for open AI API in repository environments.